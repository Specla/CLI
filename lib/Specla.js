const fs = require('fs')
const path = require('path')
const EventEmitter = require('events')
const dotenv = require('dotenv')
const express = require('express')
const cookieParser = require('cookie-parser')
const session = require('express-session')
const compression = require('compression')

const Router = require('./router')
const Controller = require('./controller')

class Specla extends EventEmitter {

  /**
   * Create new instance of the Specla app
   * @param {Object}    config
   * @return {Express}  the express app instance
   */
  constructor (config) {
    super()

    Specla.app = express()
    Specla.config = config

    // TODO find a better way to detect if the config is set in the constructor
    this.config = config

    this.rootPath = process.cwd()

    this.loadEnv()
    this.loadConfig()
    this.setupRouter()
    this.setupViewEngine()
    this.setupStaticFolder()
    this.setupCookieParser()
    this.setupSessionHandler()
    this.setupCompression()

    /**
     * Exports on isntanciation
     */
    this.app = Specla.app
    this.config = Specla.config
  }

  /**
   * Load the .env file if it exists in the project root
   * @private
   */
  loadEnv () {
    if (!fs.existsSync(path.resolve(this.rootPath, '.env'))) {
      return
    }

    dotenv.config()
  }

  /**
   * Load the config file if the config isn't specified as an option when the
   * app is instanciated
   * @private
   */
  loadConfig () {
    const configPath = path.resolve(this.rootPath, 'config.js')

    if (this.config) {
      return
    }

    if (!fs.existsSync(configPath)) {
      Specla.config = {}
      return
    }

    Specla.config = require(configPath)
  }

  /**
   * Require the routes file and let the express use the router
   * @private
   */
  setupRouter () {
    if (!Specla.config.paths) {
      Specla.config.paths = {}
    }

    const routerPath = path.resolve(
      this.rootPath,
      Specla.config.paths.routes || 'api/routes.js'
    )

    if (!fs.existsSync(routerPath)) {
      // TODO throw new Error(`Couldn't find the routes.js file`)
      return
    }

    require(routerPath)
    Specla.app.use(Specla.router)
  }

  /**
   * Setup the view engine for express, default is pug
   * @private
   */
  setupViewEngine () {
    if (!Specla.config.view) {
      Specla.config.view = {}
    }

    Specla.app.set('view engine', Specla.config.view.engine || 'pug')
    Specla.app.set('views', Specla.config.view.path || 'src/views')
  }

  /**
   * Setup the static/public folder
   * @private
   */
  setupStaticFolder () {
    if (!Specla.config.paths) {
      Specla.config.paths = {}
    }

    Specla.app.use(express.static(Specla.config.paths.static || 'static'))
  }

  /**
   * Setup the express cookie parser
   * @private
   */
  setupCookieParser () {
    Specla.app.use(cookieParser(Specla.config.key))
  }

  /**
   * Setup the session handler for the express app
   * TODO Usage of different stores
   * @private
   */
  setupSessionHandler () {
    if (!Specla.config.session) {
      Specla.config.session = {}
    }

    let options = {
      resave: false,
      secret: Specla.config.key || 'bad-key',
      saveUninitialized: true,
      cookie: {}
    }

    if (process.env.NODE_ENV === 'production') {
      Specla.app.set('trust proxy', 1)
      options.cookie.secret = true
    }

    Specla.app.use(session(options))
  }

  /**
   * Setup compression reponse body
   * @private
   */
  setupCompression () {
    Specla.app.use(compression())
  }
}

/**
 * Specla module exports
 */
Specla.app = null
Specla.config = null
Specla.router = new Router()
Specla.Controller = Controller

module.exports = Specla
