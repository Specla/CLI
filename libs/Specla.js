'use strict';

const express = require('express');
const bodyParser = require('body-parser');

const Database = require('specla-database');
const Autoloader = require('specla-autoloader');
const Router = require('specla-router');

class Specla {

  constructor(config){
    this.config = config;
    this.express = null;
    this.modules = {};
    this.models = {};
    this.events = {};

    global.Specla = this;

    this.setupProcessEvents();
    this.setupExpress();
    this.setupSpecla();
    this.setupRoutes();
  }

  /**
   * Setup process events
   * @private
   */
  setupProcessEvents(){
    process.on('exit', this.trigger.bind(this, 'exit'));
    process.on('SIGINT', () => {
      this.trigger.bind(this, 'exit');
      process.exit();
    });
  }

  /**
   * Setup express
   * @private
   */
  setupExpress(){
    this.express = express();

    this.express.use(express.static('public'));
    this.express.use(bodyParser.urlencoded({ extended: true }));
    this.express.set('view engine', this.config.view.engine);
    this.express.set('views', this.config.view.path);
  }

  /**
   * Setup Specla and its modules
   * @private
   */
  setupSpecla(){
    this.modules.db = new Database({
      host: this.config.database.host,
      port: this.config.database.port,
      database: this.config.database.database
    });

    this.modules.Model = this.modules.db.Model;

    new Autoloader(this.config.autoloader.global).global();
  }

  /**
   * Setup the router (default express router)
   * @private
   */
  setupRoutes(){
    this.modules.router = new Router({
      express: this.express
    });

    require(process.cwd()+'/api/routes');
  }

  /**
   * Register event listener
   * @param {String}    event
   * @param {Function}  callback
   * @return {Object}   this
   * @public
   */
  on(event, callback){
    if(this.events[event] === undefined){
      this.events[event] = [];
    }

    this.events[event].push(callback);

    return this;
  }

  /**
   * Remove event listener
   * @param {String}  event
   * @return {Object} this
   * @public
   */
  off(event){
    if(this.events[event] === undefined){
      return this;
    }

    delete this.events[event];
  }

  /**
   * Trigger event listener
   * @param {String}  event
   * @param {Mixed}   data
   * @return {Object} this
   * @public
   */
  trigger(event, ...data){
    if(data.length === 1){
      data = data[0];
    }

    if(data.length === 0){
      data = undefined;
    }

    if(this.events[event] === undefined){
      return this;
    }

    for(let _event of this.events[event]){
      _event.apply(data);
    }

    return this;
  }

  /**
   * Start the HTTP server from the server module
   * @param {Function} callback
   * @public
   */
  listen(callback){
    this.express.listen(this.config.server.port, () => {
      this.trigger('ready');

      if(typeof callback === 'function'){
        callback();
      }
    });
  }
}

module.exports = Specla;
