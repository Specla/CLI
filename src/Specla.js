import Config from './Config'
import defaultConfig from './defaultConfig'
import express from 'express'

export default class Specla {
  /**
   * Create a new instance of Specla
   * @param  {Object} config
   * @return {Specla}
   */
  constructor (config) {
    this.config = new Config(Object.assign({}, defaultConfig, config))
    this._express = experss(this.config.get('express'))
  }

  /**
   * Proxy the use method to express
   * @param  {Array} args
   * @public
   */
  use (...args) {
    return this._express.use(...args)
  }
}
