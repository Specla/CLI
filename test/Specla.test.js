/* eslint-env jest */
import Specla from '../src'
import setup from './utils/setup'
import cleanup from './utils/cleanup'

beforeEach(setup)
afterEach(cleanup)

test('Should configure and create new application', () => {
  const specla = new Specla()
  expect(specla.config).toMatchSnapshot()
})

test('Should expose the version from package.json as a static property', () => {
  const specla = new Specla()
  expect(Specla.version).toBe(specla.version)
})

test('Should expose the config object as a static property', () => {
  const specla = new Specla()
  expect(Specla.config).toBe(specla.config)
})

test('Should only create a server instance if the runtime is server', () => {
  const speclaCli = new Specla({ 'specla.runtime': 'cli' })
  expect(speclaCli.server).toBe(undefined)
  expect(Specla.server).toBe(null)
  Specla.reset()

  const speclaServer = new Specla()
  expect(!!speclaServer.server._express).toBe(true)
  expect(speclaServer.server).toBe(Specla.server)
})

test('Should throw error if multiple instances of Specla is created', () => {
  new Specla() // eslint-disable-line
  expect(() => new Specla()).toThrowErrorMatchingSnapshot()
})
