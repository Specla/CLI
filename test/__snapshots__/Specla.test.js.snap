// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should configure and create new application 1`] = `
Object {
  "specla": Object {
    "build": Object {
      "path": "./build",
    },
    "command": Object {
      "path": "./build/commands",
    },
    "config": Object {
      "path": "./build/config",
    },
    "src": Object {
      "path": "./src",
    },
  },
}
`;

exports[`Should throw error if multiple instances of Specla is created 1`] = `"Specla is already instanciated, use the .reset() method if you want a new instance"`;
